### PR TITLE
PWN-4857 - device + social share login logic added

### PR DESCRIPTION
**Pull request by kabramovp2p**
## Jira Ticket

https://p2pvalidator.atlassian.net/browse/PWN-4857

## Description of Work

Device + social share login logic added

### CODE REVIEW

File: app/src/main/java/org/p2p/wallet/auth/interactor/restore/SocialShareRestoreInteractor.kt [hunk 1]
```diff
@@ -1,14 +1,18 @@
...
 class SocialShareRestoreInteractor(
-    private val restoreFlowDataLocalRepository: RestoreFlowDataLocalRepository
+    private val restoreFlowDataLocalRepository: RestoreFlowDataLocalRepository,
+    private val userSignUpDetailsStorage: UserSignUpDetailsStorage,
 ) {
     fun restoreSocialShare(socialShare: String, socialShareUserId: String) {
         restoreFlowDataLocalRepository.also {
             it.socialShare = socialShare
             it.socialShareUserId = socialShareUserId
+            it.deviceShare =
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/auth/interactor/restore/SocialShareRestoreInteractor.kt:14]: why SocialShareInteractor is responsible for restoring a deviceShare too? Can you extract id?

File: app/src/main/java/org/p2p/wallet/auth/interactor/restore/UserRestoreInteractor.kt [hunk 2]
```diff
@@ -20,17 +20,50 @@ class UserRestoreInteractor(
...
     sealed interface RestoreUserWay {
+        object DevicePlusSocialShareWay : RestoreUserWay
         object SocialPlusCustomShareWay : RestoreUserWay
     }
 
     fun isUserReadyToBeRestored(): Boolean {
-        val isTwoSharesAvailable =
-            restoreFlowDataLocalRepository.socialShare != null && restoreFlowDataLocalRepository.customShare != null
-        return isTwoSharesAvailable
+        val deviceShareCount = countShare(restoreFlowDataLocalRepository.deviceShare) +
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/auth/interactor/restore/UserRestoreInteractor.kt:28]: deviceShareCount -> totalSharesCount

File: app/src/main/java/org/p2p/wallet/auth/ui/restore/common/CommonRestorePresenter.kt [hunk 3]
```diff
@@ -25,11 +28,31 @@ class CommonRestorePresenter(
...
+        when (result) {
+            is UserRestoreInteractor.RestoreUserResult.RestoreSuccessful -> {
+                restoreUserRestoreInteractor.finishAuthFlow()
+                view?.setLoadingState(isScreenLoading = false)
+                view?.navigateToPinCreate()
+            }
+            is UserRestoreInteractor.RestoreUserResult.RestoreFailed -> {
+                Timber.e(result)
+                view?.setLoadingState(isScreenLoading = false)
+                view?.showUiKitSnackBar(messageResId = R.string.error_general_message)
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/auth/ui/restore/common/CommonRestorePresenter.kt:54]: if shares not match - we show alert dialog "Shares not match"

File: app/src/main/java/org/p2p/wallet/auth/ui/smsinput/NewSmsInputPresenter.kt [hunk 4]
```diff
@@ -104,7 +104,7 @@ class NewSmsInputPresenter(
 
             restoreWalletRestoreInteractor.finishRestoreCustomShare(smsCode)
 
-            if (userRestoreInteractor.isUserReadyToBeRestored()) {
+            if (userRestoreInteractor.isUserReadyToBeRestoredByPhone()) {
```
Comment (gslevinkov) [inline app/src/main/java/org/p2p/wallet/auth/ui/smsinput/NewSmsInputPresenter.kt:107]: you can pass RestoreWay into the method and return result depending on it

userRestoreInteractor.isUserReadyToBeRestored(SocialPlusCustomShare) for example

**Review by gslevinkov — APPROVED** [review]

**Review by kuznetsov — APPROVED** [review]

File: app/src/main/java/org/p2p/wallet/auth/ui/restore/common/CommonRestorePresenter.kt (hunk 3 above)
Comment (kabramovp2p) [reply app/src/main/java/org/p2p/wallet/auth/ui/restore/common/CommonRestorePresenter.kt:54]: not needed on android side

File: app/src/main/java/org/p2p/wallet/auth/interactor/restore/UserRestoreInteractor.kt [hunk 5]
```diff
@@ -20,17 +20,50 @@ class UserRestoreInteractor(
...
+            countShare(restoreFlowDataLocalRepository.socialShare) +
+            countShare(restoreFlowDataLocalRepository.customShare)
+        return deviceShareCount > 1
+    }
+
+    fun isUserReadyToBeRestoredByPhone(): Boolean {
```
Comment (davran1001) [inline app/src/main/java/org/p2p/wallet/auth/interactor/restore/UserRestoreInteractor.kt:34]: I think we need connect this methods, and remove useless, let's see way to check isUserReadyForRestore(restoreWay: Object which contains desc of restore way)

**Review by davran1001 — APPROVED** [review]

**Pull request merged**